### PR TITLE
Templating

### PR DESCRIPTION
**Pull request by emberian**
I think these match the intention of the spec.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -31,17 +34,33 @@ pub enum Wildcard {
...
-        "[[object]]" => Some(Wildcard::Object),
-        "[[boolean]]" => Some(Wildcard::Boolean),
-        "[[array]]" => Some(Wildcard::Array),
-        "[[null]]" => Some(Wildcard::Null),
+        "[[string]]" => Some(vec![Wildcard::String]),
+        "[[number]]" => Some(vec![Wildcard::Number]),
+        "[[object]]" => Some(vec![Wildcard::Object]),
+        "[[boolean]]" => Some(vec![Wildcard::Boolean]),
+        "[[array]]" => Some(vec![Wildcard::Array]),
+        "[[null]]" => Some(vec![Wildcard::Null]),
```
Comment (jviotti) [inline src/lib.rs:77]: Why does this return a vector now? Is it because we will later implement multi-type wildcards?

File: src/schema.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+/// Schema
+pub struct Schema {
+    files: Vec<File>,
+}
+
+/// File
+pub struct File {
+    /// The filetype to generate
+    type_: String,
```
Comment (abrodersen) [inline src/schema.rs]: Wouldn't this make more sense as an enum?

Comment (jviotti) [inline src/schema.rs]: Is the trailing underscore a typo?

File: src/schema.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    /// Whether to operate in fileset mode
+    fileset: bool,
```
Comment (jviotti) [inline src/schema.rs]: Is there a way to express default values for struct members? (e.g. we could default this to `false`)

File: src/schema.rs (hunk 2 above)
Comment (abrodersen) [reply src/schema.rs]: @jviotti `type` is a reserved word.

File: src/schema.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+//! Schema
+
+#![allow(missing_docs)]
+use template;
+use std::collections::BTreeMap;
+use serde_json::Value;
+
+/// Schema
+pub struct Schema {
+    files: Vec<File>,
```
Comment (jviotti) [inline src/schema.rs]: If the schema is a vector of files, does it mean that the name of the file (which acts as a key in the JSON representation) gets lost when converting from JSON to `Schema`? If so, how would we resolve references between files?

File: src/schema.rs (hunk 3 above)
Comment (abrodersen) [reply src/schema.rs]: @jviotti You can specify that a type has a default value by implementing the `Default` trait.

File: src/schema.rs (hunk 2 above)
Comment (jviotti) [reply src/schema.rs]: Yeah, we should keep an enum of all supported file types.

Comment (jviotti) [reply src/schema.rs]: I see. Maybe we can change it to `format`?

File: src/lib.rs (hunk 1 above)
Comment (emberian) [reply src/lib.rs:77]: Yep.

Comment (emberian) [reply src/lib.rs:77]: At least, multi-type wildcards are in the schema spec :)

File: src/schema.rs (hunk 4 above)
Comment (emberian) [reply src/schema.rs]: Oops, you're right, this should be a map

File: src/schema.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        on_primary: u64,
+        /// The partition number in the logical partition table
+        logical_number: u64,
+    },
+}
+
+/// Location
+pub enum Location {
+    Independent {
+        /// The path on the filesystem where the contents while be written
```
Comment (lurch) [inline src/schema.rs:47]: while -> will ?

File: src/schema.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+/// A partition where a filesystem can be found
+pub enum Partition {
+    /// A primary partition
+    Primary(u64),
+    /// A logical partitoin
+    Logical {
+        /// The primary partition containing the logical partition table
+        on_primary: u64,
```
Comment (lurch) [inline src/schema.rs:38]: I dunno rust, but this has a maximum value of `4`, so `u64` _might_ be overkill? ;)

File: src/schema.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+/// Schema
+pub struct Schema {
+    files: BTreeMap<String, File>,
+}
+
+/// Supported output file formats
+pub enum FileFormat {
+    Ini,
+    Json,
```
Comment (lurch) [inline src/schema.rs:16]: Could add XML? :smiling_imp:

File: src/schema.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+use std::collections::BTreeMap;
+use serde_json::Value;
+
+/// Schema
+pub struct Schema {
+    files: BTreeMap<String, File>,
+}
+
+/// Supported output file formats
+pub enum FileFormat {
```
Comment (jviotti) [inline src/schema.rs:14]: Should we have some logic to transform a string to a `FileFormat`, given we'll need to do that transformation when parsing the schema?

File: src/schema.rs (hunk 6 above)
Comment (jviotti) [reply src/schema.rs:38]: It'd be so nice if we can create a subtype of an unsigned integer type constrained to `0..4`, like Ada :)

Comment (abrodersen) [reply src/schema.rs:38]: Also, isn't this all MBR specific? What about GPT?

File: src/schema.rs (hunk 8 above)
Comment (abrodersen) [reply src/schema.rs:14]: @jviotti I'm guessing that `FileFormat` will eventually implement the `Parse` trait.

File: src/schema.rs (hunk 6 above)
Comment (lurch) [reply src/schema.rs:38]: IIRC in GPT you don't have 'Logical' partitions?

**Pull request merged**